packages/meditrak-server/src/routes/userEntityPermissions/assertUserEntityPermissionPermissions.js

```diff
@@ -23,6 +27,52 @@ export const assertUserEntityPermissionPermissions = async (
   return true;
 };
 
+export const assertUserEntityPermissionEditPermissions = async (
+  accessPolicy,
+  models,
+  userEntityPermissionId,
+  updatedFields,
+) => {
+  // Check we have permission to access the record we're trying to edit
+  await assertUserEntityPermissionPermissions(accessPolicy, models, userEntityPermissionId);
+  // Check we have permission for the changes
+  await assertUserEntityPermissionUpsertPermissions(accessPolicy, models, updatedFields);
+
+  // Final check to make sure we're not editing a BES admin access permission
+  // There's nothing we could edit that couldn't be abused
+  const userEntityPermission = await models.userEntityPermission.findById(userEntityPermissionId);
+  const permissionGroup = await models.permissionGroup.findById(
+    userEntityPermission.permission_group_id,
+  );
+  if (permissionGroup.name === BES_ADMIN_PERMISSION_GROUP) {
+    throw new Error('Need BES Admin access to make this change');
+  }
+
+  return true;
+};
+
+export const assertUserEntityPermissionUpsertPermissions = async (
+  accessPolicy,
+  models,
+  { permission_group_id: permissionGroupId, entity_id: entityId },
+) => {
+  // Check we're not trying to give someone:
+  // BES admin access
+  // Access to an entity we don't have admin panel access
+  if (permissionGroupId) {
+    const permissionGroup = await models.permissionGroup.findById(permissionGroupId);
+    if (permissionGroup.name === BES_ADMIN_PERMISSION_GROUP) {
+      throw new Error('Need BES Admin access to make this change');
+    }
+  }
+  if (entityId) {
+    const entity = await models.entity.findById(entityId);
+    if (!accessPolicy.allows(entity.country_code, TUPAIA_ADMIN_PANEL_PERMISSION_GROUP)) {
+      throw new Error('Need Admin Panel access to the updated entity');
+    }
+  }
```

Comment:
All of the potential abuse cases are very nicely handled. I am a bit wary about the number of individual db calls we're adding in every time someone hits an endpoint, but we can optimise when we experience performance issues rather than making this code uglier now